src/irmin-layers/irmin_layers_intf.ml

```diff
@@ -28,6 +28,38 @@ module type S = sig
     ?recovery:bool ->
     repo ->
     unit Lwt.t
+  (** [freeze ?min ?max ?squash ?copy_in_upper ?min_upper ?recovery t] launches
+      an asynchronous freezing operation on the repo [t] to reduce the size of
+      the upper layer.
+
+      Let [o] be the set of objects (i.e. commits, nodes and contents) reachable
+      from the [max] commits and bounded by the [min] commits. During the
+      freeze, all objects in [o] are copied to the lower layer, if there is one.
+      Setting [squash] to [true] is equivalent to setting [min] to [max]. [min]
+      defaults to the empty list and [max] defaults to the head commits of the
+      repo.
+
+      Let [o'] be the set of objects reachable from the [max] commits and
+      bounded by the [min_upper] commits. When the freeze is over, if
+      [copy_in_upper] is true, then the upper layer will only contain the
+      objects of [o'], otherwise the upper layer will left be empty. [min_upper]
+      defaults to the empty list and [copy_in_upper] defaults to the repo's
+      configuration.
```

Comment:
that's a bit strange: it means that when `copy_in_upper = false` `min_upper` is discarded. That seems to suggest that the API is not optimal. Maybe `copy_in_upper` should be called `upper` and be a variant `'Empty | 'From <min-commits>`?

Reply:
The API is confusing because 
- Indeed if `copy_in_upper = false` then `min_upper` is ignored; 
- If `copy_in_upper = true` but `min_upper` is the empty list, then only the `max` are copied to the next upper - in this case `[o']` only contains the `max`. This is different from `[o]` which contains all objects reachable from `max` when `min` is empty. We should clarify this, maybe using the variant, but also in the doc.

Reply:
What about that? 
```ml
next_upper: [< `Empty | `Max_only | `Self_contained | `Min of commit list ]
```

- `` `Empty`` for `copy_in_upper=false`
- `` `Max_only`` for `copy_in_upper=true` & `max=max` & `min=max`
- `` `Self_contained`` for `copy_in_upper=true` & `max=max` & `min=[]`
- `` `Min min'`` for `copy_in_upper=true` & `max=max` & `min=min'`

(I'm not fully happy with it, especially `` `Min``)

Reply:
the ``  `Max_only `` and ``  `Self_contained `` cases are equivalent: when min is empty, only the max is copied in upper;
So if you remove the extra ``  `Self_contained `` case I think your suggestion clarifies things. I don't mind the `` `Min`` case, as it is no longer ambiguous.

Reply:
With `` `Self_contained`` I meant a 4th behaviour where we wouldn't replace `min=[]` by `min=max`. This would copy all the objects reachable from max, like the `self_contained` function would do.

Reply:
I would not add a 4th behaviour to freeze right now, as like you said, we already have the `self_contained` function for that; (and as you point out, we have to specify the range of commits for which the upper is self contained -- commits outside of that range are not copied back in upper, so in a sense, it is not self contained for commits).

Reply:
I see the difference now. I didn't get that `CopyToUpper` only had the previous upper as a source (and not the lower)